a/b testing

rajatchopra opened the pull request:
- Implements https://trello.com/c/QbXwz8wF/139-8-ingress-a-b-traffic-bleedover
 - Required quite a bit of re-architecting because a route can now have multiple services.  The 'To' field in 'route.spec' stays as is, a new field 'SecondaryServices' is added for the purpose of specifying additional services.
 - Annotations are now available in the template.
 - The template has been suitably modified to show how weights can be used against endpoints belonging to different services (only the tls edge case has been modified.. others ignore the annotations, but can be changed anytime).

Example route:

```
{
  "kind": "Route",
  "apiVersion": "v1",
  "metadata": {
    "name": "route-edge",
    "creationTimestamp": null,
    "annotations": {
         "weight.default/hello": 80,
         "weight.default/hi": 20
     }
  },
  "spec": {
    "host": "www.helloexample.com",
    "to": {
      "kind": "Service",
      "name": "hello"
    },
    "secondaryServices": [{
      "kind": "Service",
      "name": "hi"
    }],
   "tls": {
      "termination": "edge",
      "certificate": "-----BEGIN CERTIFICATE-----\nMIIDIjCCAgqgAwIBAgIBATANBgkqhkiG9w0BAQUFADCBoTELMAkGA1UEBhMCVVMx\nCzAJBgNVBAgMAlNDMRUwEwYDVQQHDAxEZWZhdWx0IENpdHkxHDAaBgNVBAoME0Rl\nZmF1bHQgQ29tcGFueSBMdGQxEDAOBgNVBAsMB1Rlc3QgQ0ExGjAYBgNVBAMMEXd3\ndy5leGFtcGxlY2EuY29tMSIwIAYJKoZIhvcNAQkBFhNleGFtcGxlQGV4YW1wbGUu\nY29tMB4XDTE1MDExMjE0MTk0MVoXDTE2MDExMjE0MTk0MVowfDEYMBYGA1UEAwwP\nd3d3LmV4YW1wbGUuY29tMQswCQYDVQQIDAJTQzELMAkGA1UEBhMCVVMxIjAgBgkq\nhkiG9w0BCQEWE2V4YW1wbGVAZXhhbXBsZS5jb20xEDAOBgNVBAoMB0V4YW1wbGUx\nEDAOBgNVBAsMB0V4YW1wbGUwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMrv\ngu6ZTTefNN7jjiZbS/xvQjyXjYMN7oVXv76jbX8gjMOmg9m0xoVZZFAE4XyQDuCm\n47VRx5Qrf/YLXmB2VtCFvB0AhXr5zSeWzPwaAPrjA4ebG+LUo24ziS8KqNxrFs1M\nmNrQUgZyQC6XIe1JHXc9t+JlL5UZyZQC1IfaJulDAgMBAAGjDTALMAkGA1UdEwQC\nMAAwDQYJKoZIhvcNAQEFBQADggEBAFCi7ZlkMnESvzlZCvv82Pq6S46AAOTPXdFd\nTMvrh12E1sdVALF1P1oYFJzG1EiZ5ezOx88fEDTW+Lxb9anw5/KJzwtWcfsupf1m\nV7J0D3qKzw5C1wjzYHh9/Pz7B1D0KthQRATQCfNf8s6bbFLaw/dmiIUhHLtIH5Qc\nyfrejTZbOSP77z8NOWir+BWWgIDDB2//3AkDIQvT20vmkZRhkqSdT7et4NmXOX/j\njhPti4b2Fie0LeuvgaOdKjCpQQNrYthZHXeVlOLRhMTSk3qUczenkKTOhvP7IS9q\n+Dzv5hqgSfvMG392KWh5f8xXfJNs4W5KLbZyl901MeReiLrPH3w=\n-----END CERTIFICATE-----",
      "key": "-----BEGIN PRIVATE KEY-----\nMIICeAIBADANBgkqhkiG9w0BAQEFAASCAmIwggJeAgEAAoGBAMrvgu6ZTTefNN7j\njiZbS/xvQjyXjYMN7oVXv76jbX8gjMOmg9m0xoVZZFAE4XyQDuCm47VRx5Qrf/YL\nXmB2VtCFvB0AhXr5zSeWzPwaAPrjA4ebG+LUo24ziS8KqNxrFs1MmNrQUgZyQC6X\nIe1JHXc9t+JlL5UZyZQC1IfaJulDAgMBAAECgYEAnxOjEj/vrLNLMZE1Q9H7PZVF\nWdP/JQVNvQ7tCpZ3ZdjxHwkvf//aQnuxS5yX2Rnf37BS/TZu+TIkK4373CfHomSx\nUTAn2FsLmOJljupgGcoeLx5K5nu7B7rY5L1NHvdpxZ4YjeISrRtEPvRakllENU5y\ngJE8c2eQOx08ZSRE4TkCQQD7dws2/FldqwdjJucYijsJVuUdoTqxP8gWL6bB251q\nelP2/a6W2elqOcWId28560jG9ZS3cuKvnmu/4LG88vZFAkEAzphrH3673oTsHN+d\nuBd5uyrlnGjWjuiMKv2TPITZcWBjB8nJDSvLneHF59MYwejNNEof2tRjgFSdImFH\nmi995wJBAMtPjW6wiqRz0i41VuT9ZgwACJBzOdvzQJfHgSD9qgFb1CU/J/hpSRIM\nkYvrXK9MbvQFvG6x4VuyT1W8mpe1LK0CQAo8VPpffhFdRpF7psXLK/XQ/0VLkG3O\nKburipLyBg/u9ZkaL0Ley5zL5dFBjTV2Qkx367Ic2b0u9AYTCcgi2DsCQQD3zZ7B\nv7BOm7MkylKokY2MduFFXU0Bxg6pfZ7q3rvg8gqhUFbaMStPRYg6myiDiW/JfLhF\nTcFT4touIo7oriFJ\n-----END PRIVATE KEY-----",
      "caCertificate": "-----BEGIN CERTIFICATE-----\nMIIEFzCCAv+gAwIBAgIJALK1iUpF2VQLMA0GCSqGSIb3DQEBBQUAMIGhMQswCQYD\nVQQGEwJVUzELMAkGA1UECAwCU0MxFTATBgNVBAcMDERlZmF1bHQgQ2l0eTEcMBoG\nA1UECgwTRGVmYXVsdCBDb21wYW55IEx0ZDEQMA4GA1UECwwHVGVzdCBDQTEaMBgG\nA1UEAwwRd3d3LmV4YW1wbGVjYS5jb20xIjAgBgkqhkiG9w0BCQEWE2V4YW1wbGVA\nZXhhbXBsZS5jb20wHhcNMTUwMTEyMTQxNTAxWhcNMjUwMTA5MTQxNTAxWjCBoTEL\nMAkGA1UEBhMCVVMxCzAJBgNVBAgMAlNDMRUwEwYDVQQHDAxEZWZhdWx0IENpdHkx\nHDAaBgNVBAoME0RlZmF1bHQgQ29tcGFueSBMdGQxEDAOBgNVBAsMB1Rlc3QgQ0Ex\nGjAYBgNVBAMMEXd3dy5leGFtcGxlY2EuY29tMSIwIAYJKoZIhvcNAQkBFhNleGFt\ncGxlQGV4YW1wbGUuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA\nw2rK1J2NMtQj0KDug7g7HRKl5jbf0QMkMKyTU1fBtZ0cCzvsF4CqV11LK4BSVWaK\nrzkaXe99IVJnH8KdOlDl5Dh/+cJ3xdkClSyeUT4zgb6CCBqg78ePp+nN11JKuJlV\nIG1qdJpB1J5O/kCLsGcTf7RS74MtqMFo96446Zvt7YaBhWPz6gDaO/TUzfrNcGLA\nEfHVXkvVWqb3gqXUztZyVex/gtP9FXQ7gxTvJml7UkmT0VAFjtZnCqmFxpLZFZ15\n+qP9O7Q2MpsGUO/4vDAuYrKBeg1ZdPSi8gwqUP2qWsGd9MIWRv3thI2903BczDc7\nr8WaIbm37vYZAS9G56E4+wIDAQABo1AwTjAdBgNVHQ4EFgQUugLrSJshOBk5TSsU\nANs4+SmJUGwwHwYDVR0jBBgwFoAUugLrSJshOBk5TSsUANs4+SmJUGwwDAYDVR0T\nBAUwAwEB/zANBgkqhkiG9w0BAQUFAAOCAQEAaMJ33zAMV4korHo5aPfayV3uHoYZ\n1ChzP3eSsF+FjoscpoNSKs91ZXZF6LquzoNezbfiihK4PYqgwVD2+O0/Ty7UjN4S\nqzFKVR4OS/6lCJ8YncxoFpTntbvjgojf1DEataKFUN196PAANc3yz8cWHF4uvjPv\nWkgFqbIjb+7D1YgglNyovXkRDlRZl0LD1OQ0ZWhd4Ge1qx8mmmanoBeYZ9+DgpFC\nj9tQAbS867yeOryNe7sEOIpXAAqK/DTu0hB6+ySsDfMo4piXCc2aA/eI2DCuw08e\nw17Dz9WnupZjVdwTKzDhFgJZMLDqn37HQnT6EemLFqbcR0VPEnfyhDtZIQ==\n-----END CERTIFICATE-----"
    }
  },
  "status": {}
}
```

@ramr @openshift/networking

liggitt commented:
> Required quite a bit of re-architecting because a route can now have multiple services.

No kidding. The impact to everything that creates a route or displays a route is significant.

smarterclayton commented on pkg/route/api/types.go:
```diff
@@ -25,9 +25,9 @@ type RouteSpec struct {
 	// be defaulted to Service.
-	To kapi.ObjectReference
+	To []kapi.ObjectReference
 
```
Don't do this - keep the fields the same internally and externally.  When we move to ingress this is going to have to merge in, and ingress has no concept of duplicate from.

smarterclayton commented on pkg/router/controller/unique_host.go:
```diff
@@ -93,100 +93,102 @@ func (p *UniqueHost) HandleRoute(eventType watch.EventType, route *routeapi.Rout
-	key := routeKey(route)
-	routeName := routeNameKey(route)
+	keys := routeKeys(route)
+	for key := range keys {
```
I don't want to do this, there is no need.  This should all stay the same.

rajatchopra replied:
Agreed, fixed in the forthcoming update

smarterclayton commented on pkg/router/template/router.go:
```diff
@@ -99,7 +100,9 @@ type templateData struct {
-	State map[string]ServiceUnit
+	State map[string](ServiceAliasConfig)
+	// the service lookup
+	ServiceUnits map[string]ServiceUnit
```
This makes it even harder to refactor to use route objects internally, which concerns me because we need to be moving in that direction.

rajatchopra replied:
It does make it harder, but we need two maps because now its now a many-many relationship between routes and services. A route can point to many services and a service can join many routes.
Two maps is the cleanest way, or we will end up designing a custom hybrid hash+graph. How about we revisit this after the PR? Second round of restructuring probably.

smarterclayton replied:
Ok, just make sure that you're covered under the same locks Ram just
added.  And make sure you don't mutate any of the units, just copy
them before setting them into the map again.

smarterclayton commented on pkg/route/api/v1/types.go:
```diff
@@ -43,6 +43,10 @@ type RouteSpec struct {
 
+	// SecondaryServices is an extension of the 'to' field. If more than one service needs to be
+	// pointed to, then use this field
+	SecondaryServices []kapi.ObjectReference `json:"secondaryServices,omitempty"`
```
The point of this field is to hold weights, so weights should be part of the API for each one.  The annotations are tangential to the point of the PR.

smarterclayton replied:
I'm not sure on the naming yet, but I'm leaning to "alternates" or "additionalTo".  We are not assuming services as target forever.

smarterclayton commented on pkg/route/api/validation/validation.go:
```diff
@@ -37,11 +37,16 @@ func ValidateRoute(route *routeapi.Route) field.ErrorList {
 	}
-	if route.Spec.To.Kind != "Service" {
-		result = append(result, field.Invalid(specPath.Child("to", "kind"), route.Spec.To.Kind, "must reference a Service"))
+	for _, svc := range route.Spec.To {
```
Fieldpath is wrong, you need to use Index for each

smarterclayton commented on pkg/route/api/validation/validation.go:
```diff
@@ -37,11 +37,16 @@ func ValidateRoute(route *routeapi.Route) field.ErrorList {
 
-	if len(route.Spec.To.Name) == 0 {
-		result = append(result, field.Required(specPath.Child("to", "name"), ""))
+	if len(route.Spec.To) == 0 {
```
Limit the number of alternates to something reasonable, like 3.  I don't want people abusing this feature until we understand it.

rajatchopra replied:
hmm... ok. The magic number should be documented well then.

smarterclayton commented on pkg/route/graph/edges.go:
```diff
@@ -18,12 +18,14 @@ const (
-	syntheticService := &kapi.Service{}
-	syntheticService.Namespace = node.Namespace
-	syntheticService.Name = node.Spec.To.Name
+	for _, svc := range node.Spec.To {
```
This is going to screw up prioritization and grouping, but I don't have a good answer right now.

smarterclayton commented on pkg/router/controller/unique_host.go:
```diff
@@ -212,9 +214,13 @@ func (p *UniqueHost) HandleNamespaces(namespaces sets.String) error {
-func routeKey(route *routeapi.Route) string {
-	return fmt.Sprintf("%s/%s", route.Namespace, route.Spec.To.Name)
+// routeKeys returns the internal router key to use for the given Route.
+func routeKeys(route *routeapi.Route) []string {
```
This should not be necessary, because route names are the key, not services.

rajatchopra replied:
ack

smarterclayton commented on pkg/router/template/types.go:
```diff
@@ -37,8 +35,15 @@ type ServiceAliasConfig struct {
 	RoutingKeyName string
+
+	// Annotations attached to this route
+	Annotations map[string]string
```
This is fine to leave in, but I want the weights on the service, not the annotations.

eparis commented:
```
Annotations are now available in the template.
```

Makes me do a friggin happy dance. I'm betting this will let us kick 90% of customer asks back at them to 'do it themselves'.  Thank you @rajatchopra

liggitt commented:
That's  helpful, I agree , but as soon as they start attaching behavior on the router to annotations, they are really really going to want to be able to put policy around who can set those annotations

eparis commented:
@liggitt don't do that.

smarterclayton commented:
Yeah, annotations are cves and injection attacks waiting to happen.
Which is why the core weights should be part of the API for now.

rajatchopra commented:
I disagree that annotations will invite policies on them. Other places, yes, but not for the router.. because a router is compiled by the admin and the template will be designed to pick only those key-values that are user set-able. e.g. weights are exposed because the admin indeed wants the user to be able to specify them. Anything else that should not be controlled by the user should not be picked up by annotations. Use env for that instead (which is also available in the template, btw).

Regarding injection attack, its a valid concern, but introducing weights as a built-in field is safe only because it performs type-checks (e.g. "weight": "80\nsh -c rm -rf `pwd`", where it should have been only an integer). And we can achieve that by exposing a type-checking regexp function to the template. e.g. https://golang.org/pkg/regexp/#Regexp.FindAllString
So we give the tools to the end-user and we can all join Eric in the happy dance.

smarterclayton commented:
Yes, as long as we make those tools available and obvious.  But we
don't get the easy out of "it's the admins problem" - we have to walk
them down that path.  Admins will ask us how to control who can do
these annotations - they've done it for everything else, and this
won't be special.

liggitt commented:
> a router is compiled by the admin and the template will be designed to pick only those key-values that are user set-able. e.g. weights are exposed because the admin indeed wants the user to be able to specify them. Anything else that should not be controlled by the user should not be picked up through annotations. Use env for that instead (which is also available in the template, btw).

env doesn't cover options that are per-route but should still be reserved for privileged users. I predict requests for "protected annotations" (for this specific use... we already have that request in other areas)

eparis commented:
@liggitt @smarterclayton entirely agree, someone will want to protect who can do what.

But I'm so happy our story today is going to be 'you can do that, just be careful'. Instead of 'you're SOL'.

I don't know the haproxy config language. What can/should/must we make easy for an admin to check?  'int' and 'string without `;` or `\n`' ? Is that enough for and admin to maybe not shoot themselves in the foot?

smarterclayton commented:
@pweil- has a list of things requested.  Someone requested regex recently
for host header matching.

On May 13, 2016, at 9:03 AM, Eric Paris <notifications@github.com> wrote:

@liggitt <https://github.com/liggitt> @smarterclayton
<https://github.com/smarterclayton> entirely agree, someone will want to
protect who can do what.

But I'm so happy our story today is going to be 'you can do that, just be
careful'. Instead of 'you're SOL'.

I don't know the haproxy config language. What can/should/must we make easy
for an admin to check? 'int' and 'string without ; or \n' ? Is that enough
for and admin to maybe not shoot themselves in the foot?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8854#issuecomment-219036476>

pweil- commented on images/router/haproxy/conf/haproxy-config.template:
```diff
@@ -206,18 +206,19 @@ backend openshift_default
   option redispatch
   option forwardfor
-  balance leastconn
+  balance roundrobin
```
Why is this changing here?

rajatchopra replied:
The weights example makes sense only when the balancing is roundrobin. So it is an example only, lets say. We can revert to the default and convey this as an example.

pweil- replied:
that makes sense and also indicates that some of these options have dependencies on other settings.  Glad you highlighted that.  The template or some sort of validation will have to account for it.

pweil- commented:
> @pweil- has a list of things requested.

Things that have been requested to customize in the past:

1.  balance options
1.  stickiness settings
1.  header info
1.  weights
1.  application cookies (ie JSESSIONID)
1.  timeouts

smarterclayton commented:
Header info definitely has security risks.  Probably timeouts and
application cookies as well.

Again, this is about making sure we understand the risks and communicate
them accurately.  We can support passing annotations.  And for A/B as a
feature weights should definitely be first class and we should define them
in terms of the generic API we want (i.e. one that would in the future
support other things beyond HAProxy, like f5 and nginx).

On Fri, May 13, 2016 at 10:22 AM, Paul Weil <notifications@github.com>
wrote:

> @pweil- <https://github.com/pweil-> has a list of things requested.
>
> Things that have been requested to customize in the past:
>
>    1. balance options
>    2. stickiness settings
>    3. header info
>    4. weights
>    5. application cookies (ie JSESSIONID)
>    6. timeouts
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8854#issuecomment-219057469>
>

rajatchopra commented:
@liggitt @smarterclayton 
I have put up a new pull request, addressing the issues listed in this one - 
One thing I have stuck with - weights are still annotations. Strict type-checking is possible now though, so weights as annotations serve as a good example.

rajatchopra commented:
Closing in favour of #9070
